### PR TITLE
Warm invoke containers for start-api and start-lambda

### PR DESCRIPTION
**Pull request by jfuss**
*Issue #, if available:*
#239

*Description of changes:*

This is the start of providing the capability of a 'warm container' invoke. This PR changes some behavior, which I think has a net positive benefit.

New behavior:
When running `sam local start-api` or `sam local start-lambda` and calling a lambda (through the api or using the aws cli/sdks), we will now spin up a container but name it based on the logical Id. After the function executes, the container will be stopped (previously it was deleted). The next invoke of that function, will be directed to the stopped container, giving a 'warm' container invoke. This avoids the (expensive) cost of creating a container. In my rough testing (on my own machine), this reduced invoke time by almost 50%; 4.1s to 1.9s for the full command and 2.5s to 1.3s for the time around all the docker calls (create, invoke, delete containers).

Caveats:
* Since we name the container based on the logical id (function_name in the code), we can only have one container up at a time.
* We delete containers in a finally block (not trilled about this), so any failure will crash the cli.
* I imported some docker stuff into the services files, which leaks the docker into those areas (in my opinion, it doesn't belong there).

*Checklist:*

- [ ] Write Design Document ([Do I need to write a design document?](https://github.com/awslabs/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.rst#design-document))
- [ ] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [ ] `make pr` passes
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by j** [comment]
Unless I did it wrong (probably), I didn't see a speed boost from a node.js app.  But it could be that I'm mounting my whole project directory?  Not sure.

**Comment by jfuss** [comment]
@j The first invoke will still be 'slow'. I didn't add starting up all containers on command run. So we have the same behavior as before on the first invoke, but invokes after should be much faster. I did this to make the start up cost not huge and unless you are constantly changing your template, you can 'reload' code changes by invoking again. Let me know if you have further questions or are still not seeing any speed boost.

File: samcli/local/docker/manager.py [hunk 1]
```diff
@@ -99,7 +99,15 @@ def run(self, container, input_data=None, warm=False):
...
+
+        :param samcli.local.docker.container.Container container: Container to stop
+        """
+        container.stop()
+
+    def delete(self, container):
+        """
+        Delete the container
 
         :param samcli.local.docker.container.Container container: Container to stop
```
Comment (dsanders11) [inline samcli/local/docker/manager.py]: Container to stop => Container to delete

**Comment by lopun** [comment]
Is there any working progress on this PR? I'm really looking forward for this feature to be deployed :smile:

**Comment by yang-wei** [comment]
Is there any working progress on this PR? I'm really looking forward for this feature to be deployed 😄

**Comment by c2tarun** [comment]
Hi,

We are working on other items currently. We do plan to work on this feature (faster invokes) but do not have an eta now. This PR is left open as a base to when this worked gets picked again.

Thanks

**Comment by jfuss** [comment]
Closing in favor of the newer pr: #2383

**Pull request closed without merging**